Advise users about potential accessibility failings in their posts - RFC #58

aitchiss opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR implements RFC https://github.com/forem/rfcs/pull/58 - Advise users about potentially accessibility failings in their posts. 

### The accessibility issues it detects are:

- If a level one heading has been used anywhere (we turn the post title into a level one heading,  and ideally there should only be one on a given page)
- If heading levels have been skipped, for example going from a level 2 heading directly to a level 4 (instead of 3)
- If image alt text has been left blank, i.e. `![](image-url.png)`
- If image alt text has been left as the default we provide after upload, i.e. `![alt text](image-url.png)`

### The feature works as follows:

- When a user clicks the "Preview" button in the post editor, we asynchronously lint the markdown for issues using [markdownlint](https://github.com/DavidAnson/markdownlint)
- If there are no errors to display in the preview window, and there are some accessibility suggestions from the linting, then we show an "info" style notice in the preview window
- The notice shows a maximum of 3 suggestions (to keep the UI manageable / not overwhelming)
- The notice favours image alt text issues over heading errors, since they are more impactful
- The notice should suggest a change, specify which image/heading is referred to, and link to the relevant section of the editor guide


## Related Tickets & Documents

https://github.com/forem/rfcs/pull/58

## QA Instructions, Screenshots, Recordings

https://user-images.githubusercontent.com/20773163/117985827-f3803580-b330-11eb-95c0-e4436e7f82bd.mp4


### Checking the main behaviours:

- We should test this in both versions of the editor
- Experiment with adding the various errors mentioned above
- Check that where more than 3 suggestions exist, image suggestions are favoured
- Check that the message displayed always points to the text where the error occurred (i.e. the specific markdown for the image or heading)
- Check that if you go back and implement a suggestion, then it disappears when you return to the Preview tab
- Check that if you haven't made any of the above errors, no notice is shown

### Checking edge cases

-  We should check using the different methods of adding images (typing manually and linking to an external image, uploading via clicking "upload image", drag and drop, copy/paste)
- Check that if you add an image in the middle of a line - e.g. `some text ![](image-url.png) some other text` - the correct text is pulled out in the message in the Preview notice
- Trigger an error in the markdown by adding e.g. `{%tag %}` in the text, and make sure that when the error message is shown, the accessibility suggestions notice isn't

### UI accessibility concerns?

-  The notice has an `aria-live="polite"` to make sure it's announced when it's populated - we can check this with a screen reader
- The "Learn more" links next to each suggestion have a more informative `aria-label` ("Learn more about accessible images" / "Learn more about accessible headings") - we can also check that with a screen reader

## Added tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [X] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post


## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

![it's a lot to process](https://media.giphy.com/media/XdCvqP8xRjgSdFhAbZ/giphy.gif)

aitchiss commented on app/javascript/article-form/articleForm.jsx:
```diff
@@ -160,7 +187,41 @@ export class ArticleForm extends Component {
+  };
+
+  fetchMarkdownLint = async () => {
+    if (!window.markdownlint) {
+      const markdownItScript = document.createElement('script');
+      markdownItScript.setAttribute('src', '/assets/markdown-it.min.js');
+      document.body.appendChild(markdownItScript);
+
+      // The markdownlint script needs the first script to have finished loading first
+      markdownItScript.addEventListener('load', () => {
+        const markdownLintScript = document.createElement('script');
+        markdownLintScript.setAttribute(
+          'src',
+          '/assets/markdownlint-browser.min.js',
+        );
+        document.body.appendChild(markdownLintScript);
+
+        markdownLintScript.addEventListener('load', this.lintMarkdown);
+      });
+    } else {
+      this.lintMarkdown();
+    }
```
The markdownlint lib is [designed to be used in the browser by adding as script tags](https://github.com/DavidAnson/markdownlint#browser) rather than importing from a module. @nickytonline if you can suggest a better way of loading it in - let me know 🙏 

This loads it in once asynchronously (when the Preview tab is clicked), and from that point on the `markdownlint` method is available on the window object. This whole method is called async so that linting the markdown doesn't block any other preview functionality.

rhymes replied:
I installed `markdown-it` with `yarn add markdown-it` and I found `markdown-it.min.js` in the `node_modules/markdown-it/dist` folder so I think we can require that directly from there in the editor. It's the only dependency `markdownlint` declares so it'll be there :)  

`markdownlint` instead doesn't seem to offer a bundled browser version :( maybe we can put it in a `app/javascript/vendor` so we know it's a third party script.

It's not a small library but I don't think we have an alternative: https://bundlephobia.com/result?p=markdownlint@0.23.1

rhymes commented on app/javascript/utilities/markdown/markdownLintCustomRules.js:
```diff
@@ -0,0 +1,146 @@
+          ) {
+            onError({
+              lineNumber: contentChild.lineNumber,
+              detail: '/p/editor_guide#alt-text-for-images',
```
I'm thinking of ways not to hardcode this but I can't find any easy ones

rhymes commented on app/javascript/article-form/components/AccessibilitySuggestions.jsx:
```diff
@@ -0,0 +1,94 @@
+              <span className="fs-s">
+                {' '}
+                <a
+                  href={lintError.errorDetail}
```
I think these links should open a in new tab. If I click while I'm editing I get redirected to the linked page and I'm not 100% sure we should send people away from the editor

aitchiss replied:
done!

rhymes reviewed (commented):
Great stuff @aitchiss! I left some thoughts inline.

One thing: we support the `<img>` HTML tag inside markdown (and other HTML tags as well), so we might want to consider adding a rule for alt text for that as well.

nickytonline commented:
> Great stuff @aitchiss! I left some thoughts inline.
> 
> One thing: we support the `<img>` HTML tag inside markdown (and other HTML tags as well), so we might want to consider adding a rule for alt text for that as well.

@rhymes, they do offer an npm package, but it is for the backend, i.e. node.js is required because it requires the Node.js fs import. We came across this when [I paired with a community member last October in regards to linting markdown](https://dev.to/devteam/pairing-with-community-member-eliot-sanford-f7a). We can't dynamically import it as far as I know, but we can inject the script when the editor loads/has loaded.

lisasy reviewed (approved):
Overall this looks good to me. I think that the constraints we're setting over the number of tips to show is good. The language is clear. One recommendation I want to explore is making each "Learn more" link open into a new tab to not divert people's current flow from creating their post. If it's doable by making `target="_blank"`, is this appropriate?

![Image 2021-05-28 at 9 08 44 AM](https://user-images.githubusercontent.com/816402/120012454-62e26000-bf94-11eb-9566-f9706de18bf6.jpg)

Overall I think this is exciting. As we make incremental improvements to the composer experience, I can see how we may begin branding this message as a message coming from Forem's Sloan, where we could add the illustration and update the copy to reflect our brand voice. But that's not for this PR ☺️

aitchiss commented:
@lisasy @rhymes - I've updated the "Learn more" links so that they open in a new tab, and added an icon to reflect that 🙂 

<img width="813" alt="screenshot showing the learn more links now have a 'new window/tab' icon to the right" src="https://user-images.githubusercontent.com/20773163/120303969-3fd2eb80-c2c7-11eb-99af-9d9c8337e22c.png">

nickytonline reviewed (approved):
Thanks for this @aitchiss! 🛳
